### PR TITLE
(feat) O3-1831: Registration: Support person attribute of type Location

### PR DESCRIPTION
**Pull request by usamaidrsk**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR add support for person attribute type Location, on patient registration.

test config
```typescript
{
  "@openmrs/esm-patient-registration-app": {
    "fieldDefinitions": [
      {
        "id": "Health Center",
        "label": "Heath Center",
        "type": "person attribute",
        "uuid": "8d87236c-c2cc-11de-8d13-0010c6dffd0f"
      }
    ],
    "sections": [
      "demographics",
      "contact",
      "relationships",
      "extraInfo"
    ],
    "sectionDefinitions": [
      {
        "id": "extraInfo",
        "name": "Extra info",
        "fields": [
          "Health Center"
        ]
      }
    ]
  }
}
``` 

## Screenshots

https://github.com/user-attachments/assets/d07bfb8b-b1ed-40e6-b216-d1c55f12aa8e

## Related Issue

[O3-1831](https://openmrs.atlassian.net/browse/O3-1831
)
## Other

[O3-1831]: https://openmrs.atlassian.net/browse/O3-1831?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by brandones** [comment]
@usamaidrsk Seems like this is very close. Can you wrap it up and I'll review?

**Comment by brandones** [comment]
Ping @usamaidrsk ? It's still failing CI and still a draft.

**Comment by usamaidrsk** [comment]
This work is being blocked by [TRUNK-6246](https://openmrs.atlassian.net/browse/TRUNK-6246)
cc @brandones cc @ibacher

**Comment by brandones** [comment]
@usamaidrsk Looks like the blocking issue has not been resolved, this is still in draft, and CI is failing. Why did you review request me and Ian?

File: packages/esm-patient-registration-app/src/config-schema.ts [hunk 1]
```diff
@@ -167,6 +168,12 @@ export const esmPatientRegistrationSchema = {
           _description: 'Optional RegEx for testing the validity of the input.',
         },
       },
+      locationTag: {
+        _type: Type.String,
+        _default: null,
+        _description:
+          'For locations questions only. A concept which has the possible responses either as answers or as set members.',
```
Comment (mseaton) [inline packages/esm-patient-registration-app/src/config-schema.ts:175]: This description doesn't seem to make sense?

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  personAttributeType,
+  id,
+  label,
+  locationTag,
+  required,
+}: LocationPersonAttributeFieldProps) {
+  const { t } = useTranslation();
+  const [field, meta, { setValue }] = useField(`attributes.${personAttributeType.uuid}`);
+  const [searchQuery, setSearchQuery] = useState<string>('');
+  const { locations } = useLocations(locationTag || null, 5, searchQuery);
```
Comment (mseaton) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx:28]: What does this number 5 do?  Is that right?

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  const locationOptions = useMemo(() => {
+    return locations.map(({ resource: { id, name } }) => ({ value: id, label: name }));
+  }, [locations]);
+
+  useEffect(() => {
+    if (meta?.value?.uuid) {
+      setValue(meta.value.uuid);
```
Comment (mseaton) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx:36]: What does this do?  Is this intended to be a default value?  Can we document this and comment it?

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import { useCallback, useMemo } from 'react';
+import { type FetchResponse, fhirBaseUrl, openmrsFetch, useDebounce } from '@openmrs/esm-framework';
+import { type LocationEntry, type LocationResponse } from '@openmrs/esm-service-queues-app/src/types';
+import useSWR from 'swr';
+interface IUseLocations {
```
Comment (mseaton) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx:5]: I don't believe this adheres to our conventional coding standards for interfaces.  Don't prefix with `I` as a starter, not sure about the verbage in general...

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  locations: Array<LocationEntry>;
+  isLoading: boolean;
+  loadingNewData: boolean;
+  fetchLocationByUuid: (uuid: string) => void;
+}
+
+export function useLocations(locationTag: string | null, count: number = 0, searchQuery: string = ''): IUseLocations {
```
Comment (mseaton) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx:12]: I don't love the `count` here, but even if we keep it, defaulting to 0 seems wrong.

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  const debouncedSearchQuery = useDebounce(searchQuery);
+
+  const fetchLocationByUuid = useCallback((uuid: string) => {
+    const { data } = useSWR<FetchResponse<LocationResponse>, Error>(`${fhirBaseUrl}/Location/${uuid}`, openmrsFetch);
+  }, []);
```
Comment (ibacher) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx:17]: Don't use `useSWR()`. It's a violation of the [rules of hooks](https://react.dev/reference/rules/rules-of-hooks).

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx (hunk 5 above)
Comment (brandones) [reply packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx:12]: `count` is not even being used in the function...

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx (hunk 4 above)
Comment (brandones) [reply packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx:5]:
```suggestion
interface UseLocationsResult {
```

**Comment by brandones** [comment]
Ping @usamaidrsk , it's been a month...

**Comment by brandones** [comment]
Ping again, @usamaidrsk

**Comment by usamaidrsk** [comment]
> Ping again, @usamaidrsk

Hi @brandones 
I am following up on this [PR](https://github.com/openmrs/openmrs-module-webservices.rest/pull/620) to be merged.
Thanks cc @mherman22 @ibacher

**Comment by usamaidrsk** [comment]
Hello @brandones, this PR is now ready for review
Thanks

File: packages/esm-patient-registration-app/src/config-schema.ts [hunk 7]
```diff
@@ -153,8 +154,8 @@ export const esmPatientRegistrationSchema = {
       },
       type: {
         _type: Type.String,
-        _description: "How this field's data will be stored—a person attribute or an obs.",
-        _validators: [validators.oneOf(['person attribute', 'obs'])],
+        _description: "How this field's data will be stored—a person attribute, an obs or an address.",
+        _validators: [validators.oneOf(['person attribute', 'obs', 'address'])],
```
Comment (brandones) [inline packages/esm-patient-registration-app/src/config-schema.ts:158]: I don't think this ticket has anything to do with addresses.

```suggestion
        _description: "How this field's data will be stored—a person attribute or an obs.",
        _validators: [validators.oneOf(['person attribute', 'obs'])],
```

File: packages/esm-patient-registration-app/src/config-schema.ts [hunk 8]
```diff
@@ -183,6 +184,11 @@ export const esmPatientRegistrationSchema = {
           _description: 'Optional RegEx for testing the validity of the input.',
         },
       },
+      locationTag: {
+        _type: Type.String,
+        _default: null,
+        _description: 'For locations questions only. A tag used to filter and categorize locations.',
```
Comment (brandones) [inline packages/esm-patient-registration-app/src/config-schema.ts:190]:
```suggestion
        _description: 'Only for fields with "person attribute" type `org.openmrs.Location`. This filters the list of location options in the dropdown based on their location tag. By default, all locations are shown.',
```

Please check AI-generated text before committing. The description as written makes no sense in this context.

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/person-attribute-field.component.tsx [hunk 9]
```diff
@@ -41,6 +42,15 @@ export function PersonAttributeField({ fieldDefinition }: PersonAttributeFieldPr
             required={fieldDefinition.validation?.required ?? false}
           />
         );
+      case 'org.openmrs.Location':
+        return (
+          <LocationPersonAttributeField
+            personAttributeType={personAttributeType}
+            locationTag={fieldDefinition.locationTag}
+            label={fieldDefinition.label}
+            id={fieldDefinition?.id}
```
Comment (brandones) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/person-attribute-field.component.tsx:51]: Please include the `required` parameter, as with the above field types.

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import { useTranslation } from 'react-i18next';
+
+export interface LocationPersonAttributeFieldProps {
+  id: string;
+  personAttributeType: PersonAttributeTypeResponse;
+  validationRegex?: string;
+  label?: string;
+  locationTag: string;
+  required?: boolean;
+}
```
Comment (brandones) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx:17]:
```suggestion
export interface LocationPersonAttributeFieldProps {
  id: string;
  personAttributeType: PersonAttributeTypeResponse;
  label?: string;
  locationTag: string;
  required?: boolean;
}
```

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            return (
+              <ComboBox
+                id={id}
+                name={`person-attribute-${personAttributeType.uuid}`}
+                titleText={label}
+                items={locationOptions}
+                placeholder={t('searchLocationPersonAttribute', 'Search location')}
+                onInputChange={onInputChange}
+                required={required}
+                onChange={handleSelect}
```
Comment (brandones) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx:83]: Please make your function names at least be consistent within a file. `onInputChange` and `onChange` or `handleInputChange` and `handleSelect`. Pick one.

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  // Callback for when updating the combobox input
+  const onInputChange = useCallback(
+    (value: string | null) => {
+      if (value) {
+        // If the value exists in the locationOptions (i.e. a label matches the input), exit the function
+        if (locationOptions.find(({ label }) => label === value)) return;
+        // If the input is a new value, set the search query
+        setSearchQuery(value);
+        // Clear the current selected value since the input doesn't match any existing options
+        setValue(null);
```
Comment (brandones) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.component.tsx:55]: I would feel more confident about this if you'd posted a video showing you testing a few cases—starting to type a valid location name, typing nonsense, typing a location name exactly, selecting a location, changing the input text after having selected a location. I think this code makes sense but I don't have 100% confidence in it.

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+export function useLocations(locationTag: string | null, searchQuery: string = ''): UseLocationsResult {
+  const debouncedSearchQuery = useDebounce(searchQuery);
+
+  const constructUrl = useMemo(() => {
+    let url = `${fhirBaseUrl}/Location?`;
+    let urlSearchParameters = new URLSearchParams();
+    urlSearchParameters.append('_summary', 'data');
+
+    if (!debouncedSearchQuery) {
+      urlSearchParameters.append('_count', '5');
```
Comment (brandones) [inline packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx:21]: The screenshot you posted shows more than 5 locations, with no search query. I guess the screenshot is out of date? Please update it. Ideally with a video.

Also, why such a small number?

**Review by brandones — CHANGES_REQUESTED** [review]

File: packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx (hunk 13 above)
Comment (usamaidrsk) [reply packages/esm-patient-registration-app/src/patient-registration/field/person-attributes/location-person-attribute-field.resource.tsx:21]: I have added the video on the description, and for the count being 5, I was trying to eliminate the scroll on the select dropdown, but I can change that to a number you suggest.
Thanks

**Comment by brandones** [comment]
Thanks @usamaidrsk . The code looks much better, and the video is helpful. Unfortunately the UX seems quite bad, based on your video. You type "out" and "Outpatient clinic" disappears, you see all options disappear, and then nothing comes up, etc.

As a starting point for improving it, I think the list of options should not be cleared until results from the search have been returned by the server. It might also be nice to put a loading spinner on the right side of the input box while new results are loading, to indicate that the user's input to the search input box is being responded to.

**Comment by usamaidrsk** [comment]
Hi @brandones please take a look at this.
Thanks
cc @vasharma05 
https://github.com/user-attachments/assets/4b9e6ff0-ef00-41c1-b754-ab981752e1fa

**Comment by brandones** [comment]
Much better, thanks @usamaidrsk . But having the loading indicator push the dropdown downward creates layout shift, which is a bad UX. Please adjust the loading indicator so that it doesn't create layout shift. Maybe it could float with absolute positioning toward the right side of the input box or something.

**Comment by ibacher** [comment]
@usamaidrsk We should be using a [Dropdown](https://react.carbondesignsystem.com/?path=/docs/components-dropdown--overview) here which has a skeleton state to indicate that it's loading.

**Comment by usamaidrsk** [comment]
> @usamaidrsk We should be using a [Dropdown](https://react.carbondesignsystem.com/?path=/docs/components-dropdown--overview) here which has a skeleton state to indicate that it's loading.

Thanks @ibacher 
I believe this might be a slightly different use case. In this scenario, the loading indicator is shown while the user is actively searching/typing. Replacing the entire input with a a skeleton loader could potentially be confusing in this context.

cc @brandones

**Comment by ibacher** [comment]
I see. I'd probably opt for a loading indicator inside the combo box on its right edge.

**Comment by usamaidrsk** [comment]
> I see. I'd probably opt for a loading indicator inside the combo box on its right edge.

Yeah, that sounds good. Could we have done something similar, that I can refer too.
Thanks

**Comment by brandones** [comment]
@usamaidrsk We haven't, but it's just CSS, should be pretty straightforward. Adding something to the right side of an input box (e.g. an eye icon, an x icon) is a very common thing to do, I'm sure there's plenty of info on Google/SO/etc.

**Comment by usamaidrsk** [comment]
> @usamaidrsk We haven't, but it's just CSS, should be pretty straightforward. Adding something to the right side of an input box (e.g. an eye icon, an x icon) is a very common thing to do, I'm sure there's plenty of info on Google/SO/etc.

Hi @brandones, I have improved it, see this here!
https://github.com/user-attachments/assets/570ed2c3-dbe4-4320-abbd-328e7491eddd

Thanks

**Comment by brandones** [comment]
That looks really nice @usamaidrsk , great work. Hopefully the UX reasoning is clear to you, for future work.

**Review by brandones — APPROVED** [review]
LGTM

**Comment by brandones** [comment]
I still think it might be nicer to have a higher limit for the number of locations to load. Feel free to merge this as it is, or to up the limit. Since loading takes so long it might be nice to minimize how many times the user has to wait for that loading to happen in order to find the location they want to select, which I figure would be accomplished by loading a longer list of locations each time.

**Comment by brandones** [comment]
(please post a video if you do change it)

**Comment by usamaidrsk** [comment]
> (please post a video if you do change it)

I have increased the number of loactions loaded on searching from 5 to 10. Here is the video
https://github.com/user-attachments/assets/89da941a-a4dd-4f9d-a4e9-4d9054b269cf

**Comment by brandones** [comment]
Yeah I think that makes more sense to me. Thank you @usamaidrsk . And if someone wants to change that number again in the future it's easy enough.

**Comment by usamaidrsk** [comment]
> Hopefully the UX reasoning is clear to you, for future work.

Thanks @brandones , it was great work here!
I will merge this now

**Pull request merged**